File: Java/243-ShortestWordDistance.java
public class Solution {
    //brute force, O(n^2)
    public int shortestDistanceBF(String[] words, String word1, String word2) {
        int minDistance = words.length;
        for (int i = 0; i < words.length; i++) {
            if (words[i].equals(word1)) {
                for (int j = 0; j < words.length; j++) {
                    if (words[j].equals(word2)) {
                        minDistance = Math.min(minDistance, Math.abs(i - j));
                    }
                }
            }
        }
        return minDistance;
    }

    //two pointers, one pass O(n)
    public int shortestDistance(String[] words, String word1, String word2) {
        if (words.length == 0 || words == null)
            return 0;
        int len = words.length;
        int res = len;
        int p1 = -len, p2 = -len;
        for (int i = 0; i < len; i++){
            if (words[i].equals(word1)){
                p1 = i;
                res = Math.min(res, Math.abs(i - p2));//also works without abs
            }else if (words[i].equals(word2)){
                p2 = i;
                res = Math.min(res, Math.abs(i - p1));
            }
        }
        return res;
    }
}

File: cpp/243-ShortestWordDistance.cpp
class Solution {
public:
    int shortestDistance(vector<string>& words, string word1, string word2) {
        int p1 = -1, p2 = -1, minD = INT_MAX;
        for (int i = 0; i < words.size(); i++) {
            if (words[i] == word1) {
                p1 = i;
                if (p2 != -1)
                    minD = min(i - p2, minD);
            }
            if (words[i] == word2) {
                p2 = i;
                if (p1 != -1)
                    minD = min(i - p1, minD);
            }
        }
        return minD == INT_MAX ? -1 : minD;
    }
};
